### PR TITLE
Require email confirmation before creating the first developer API key

### LINKED ISSUE #11635: Require email confirmation for first set of API keys
From the [Account Submission Restrictions PRD](https://docs.google.com/document/d/1FfFcydrcm_TC4sQHXWQfJg2_11RvvYUANDy0hqMlH7U/edit#).

The first time a developer goes to the [API Keys page on AMO](https://addons.mozilla.org/developers/addon/api/key/), the `Generate new credentials` button shouldn't create them immediately. Instead, the following should happen:

* The page displays a message saying "A confirmation link will be sent to your email address. After confirmation you will find your API keys on this page."
* An email is sent with a temporary confirmation link. It should be possible to introduce a delay in sending this email, to protect against spamming.
* The link should point back to the API Keys page, where the keys should now be shown.

After confirmation, re-generating keys shouldn't require confirmation again. However, it should be possible to reset the confirmation status of any or all developers in order to require confirmation again. By "possible" I just mean it should be visible in the user admin and it should be easy to ask ops to change this flag for a group of users if needed.

### PR DESCRIPTION
**Pull request by diox**
Fixes #11635

### CODE REVIEW

File: src/olympia/api/models.py [hunk 1]
```diff
@@ -101,3 +106,38 @@ def generate_secret(byte_length):
...
+
+    def send_confirmation_email(self):
+        context = {
+            'api_key_confirmation_link': (
+                absolutify(reverse('devhub.api_key')) + f'?token={self.token}'
+            ),
+            'domain': settings.DOMAIN,
+        }
+        return send_mail_jinja(
+            'Confirmation for developer API keys',
```
Comment (EnTeQuAk) [inline src/olympia/api/models.py]: shouldn't that string be translated?

Comment (diox) [reply src/olympia/api/models.py]: It should yes, good catch.

File: src/olympia/lib/settings_base.py [hunk 2]
```diff
@@ -279,6 +279,7 @@ def get_db_config(environ_var, atomic_requests=True):
     r'^devhub\/email\/revoked-key-email.ltxt',
     r'^devhub\/email\/new-key-email.ltxt',
     r'^devhub\/email\/submission_api_key_revocation.txt',
+    r'^devhub\/email\/api_key_confirmation.ltxt',
```
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/lib/settings_base.py]: It's older than both of us :) I think the idea was to differentiate text templates from plain text files. There are a few outliers though.

File: src/olympia/lib/settings_base.py [hunk 3]
```diff
@@ -1649,6 +1650,11 @@ def read_only_mode(env):
 # set the expiration any longer than this.
 MAX_APIKEY_JWT_AUTH_TOKEN_LIFETIME = 5 * 60
 
+# Time in seconds before the email containing the link allowing developers to
+# see their api keys the first time they request one is sent. A value of None
+# means it's sent instantaneously.
+API_KEY_CONFIRMATION_DELAY = None
```
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/lib/settings_base.py]: That was a request for product to be able to do that if necessary with just a redeploy.

File: src/olympia/devhub/templates/devhub/email/api_key_confirmation.ltxt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{% load i18n %}{# L10n: This is an email. Whitespace matters #}{% blocktrans %}Hello,
+
+You're receiving this message because you requested new developer API keys for
+{{ domain }}. To finish this process, please click on the link below:
+
+{{ api_key_confirmation_link }}
+
+If you didn't request this, please disregard this email or notify amo-admins@mozilla.com.
```
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/devhub/templates/devhub/email/api_key_confirmation.ltxt]: Yeah maybe we can tweak that sentence to `If you didn't request this, please notify`. I'll ask in the issue as the copy has been validated already.

File: src/olympia/api/models.py (hunk 1 above)
Comment (diox) [reply src/olympia/api/models.py]: c5da7095f15a78ed9c11ac05ec65eeacd7a87eaf

File: src/olympia/devhub/templates/devhub/email/api_key_confirmation.ltxt (hunk 4 above)
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/devhub/templates/devhub/email/api_key_confirmation.ltxt]: Made this simple tweak in a88deb1aea2c9bc7c8b7a0b3a416cb9385064896

File: src/olympia/migrations/1129-add-apikeyconfirmation.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE `api_apikeyconfirmation` (
+    `created` datetime(6) NOT NULL,
+    `modified` datetime(6) NOT NULL,
+    `user_id` integer NOT NULL PRIMARY KEY,
```
Comment (EnTeQuAk) [inline src/olympia/migrations/1129-add-apikeyconfirmation.sql]: I'm confused by historical migrations, some mention `user_id` to be unsigned but newer ones don't do that… so I think we should be safe here :thinking:

Comment (diox) [reply src/olympia/migrations/1129-add-apikeyconfirmation.sql]: Yeah it shouldn't be `UNSIGNED` - it's not in dev/stage/prod schema at least, and we don't use our `PositiveAutoField` on `UserProfile`.

**Pull request merged**